Scripts/Data/Collection/GameObjectSet.cs

```diff
@@ -0,0 +1,129 @@
+        public class UnityEvent : UnityEvent<GameObject>
+        {
+        }
+
+        /// <summary>
+        /// Emitted when a searched element is found.
+        /// </summary>
+        public UnityEvent ElementFound = new UnityEvent();
+        /// <summary>
+        /// Emitted when a searched element is not found.
+        /// </summary>
+        public UnityEvent ElementNotFound = new UnityEvent();
+        /// <summary>
+        /// Emitted when the collection has the last element removed from it and becomes empty.
+        /// </summary>
+        public UnityEvent BecameEmpty = new UnityEvent();
+        /// <summary>
+        /// Emitted when the collection has the first element added to it and becomes populated.
+        /// </summary>
+        public UnityEvent BecamePopulated = new UnityEvent();
+
+        /// <summary>
+        /// The elements currently in the set.
+        /// </summary>
+        public HashSet<GameObject> Elements
+        {
+            get;
+            protected set;
+        } = new HashSet<GameObject>();
+
+
+        /// <summary>
+        /// Adds the given element to the set.
+        /// </summary>
+        /// <param name="element">The element to add.</param>
+        public virtual void AddElement(GameObject element)
+        {
+            if (!isActiveAndEnabled || element == null)
+            {
+                return;
+            }
+
+            if (Elements.Count == 0)
+            {
+                BecamePopulated?.Invoke(element);
+            }
+            Elements.Add(element);
+        }
+
+        /// <summary>
+        /// Removes the given element from the set.
+        /// </summary>
+        /// <param name="element">The element to remove.</param>
+        public virtual void RemoveElement(GameObject element)
+        {
+            if (!isActiveAndEnabled)
+            {
+                return;
+            }
+
+            if (Elements.Remove(element) && Elements.Count == 0)
+            {
+                BecameEmpty?.Invoke(element);
+            }
+        }
+
+        /// <summary>
+        /// Clears all elements from the set.
+        /// </summary>
+        public virtual void ClearElements()
+        {
+            if (!isActiveAndEnabled)
+            {
+                return;
+            }
+
+            Elements.Clear();
+        }
+
+        /// <summary>
+        /// Determines if the given element is contained within the set and emits the appropriate event based on the result.
+        /// </summary>
+        /// <param name="element">The element to search for.</param>
+        /// <returns><see langword="true"/> if the element is found.</returns>
+        public virtual bool Contains(GameObject element)
+        {
+            if (!isActiveAndEnabled)
+            {
+                return false;
+            }
+
+            bool result = Elements.Contains(element);
+            if (result)
+            {
+                ElementFound?.Invoke(element);

```

Comment:
These events are strange as there is no connection between the caller of the method and the receiver of the event.

I assume this is needed for wiring things up in the Unity inspector and we can't instead switch those to be done in a script just utilizing the return value of the method?

Reply:
Not sure what you mean here.

The event just rethrows the thing that you gave it so you can use it for chaining.

E.g.

GameObject -> is it in the set? -> if yes then do x with GameObject if no then do y with GameObject

Reply:
1. Three things listen for the event.
2. _Someone_ asks for whether a given object is contained.
3. All listeners get an answer despite not even being the one that asked.

Reply:
You wire up your events in the inspector to be things that do something when that event is fired.

So the usage is

Check set contains ObjectA
If it does contain it then do x,y,z (the three listeners)

The listener doesn't do the asking, the listener is just set up to respond to the question